jxl/src/render/simple_pipeline.rs

```diff
@@ -0,0 +1,547 @@
+    fn new(num_channels: usize, size: (usize, usize), log_group_size: usize) -> Self {
+        SimpleRenderPipelineBuilder {
+            pipeline: SimpleRenderPipeline {
+                channel_info: vec![vec![
+                    ChannelInfo {
+                        ty: None,
+                        downsample: (0, 0)
+                    };
+                    num_channels
+                ]],
+                input_size: size,
+                log_group_size,
+                xgroups: size.0.shrc(log_group_size),
+                stages: vec![],
+                group_ready_passes: vec![
+                    0;
+                    size.0.shrc(log_group_size) * size.1.shrc(log_group_size)
+                ],
+                completed_passes: 0,
+                input_buffers: vec![],
+            },
+            can_shift: true,
+        }
+    }
+    fn add_stage<Stage: RenderPipelineStage>(mut self, stage: Stage) -> Result<Self> {
+        let current_info = self.pipeline.channel_info.last().unwrap().clone();
+        let mut after_info = vec![];
+        for (c, info) in current_info.iter().enumerate() {
+            if !stage.uses_channel(c) {
+                after_info.push(ChannelInfo {
+                    ty: info.ty,
+                    downsample: (0, 0),
+                });
+            } else {
+                if let Some(ty) = info.ty {
+                    if ty != Stage::Type::INPUT_TYPE {
+                        return Err(Error::PipelineChannelTypeMismatch(
+                            stage.name(),
+                            c,
+                            Stage::Type::INPUT_TYPE,
+                            ty,
+                        ));
+                    }
+                }
+                after_info.push(ChannelInfo {
+                    ty: Some(Stage::Type::OUTPUT_TYPE.unwrap_or(Stage::Type::INPUT_TYPE)),
+                    downsample: Stage::Type::SHIFT,
+                });
+            }
+        }
+        if !self.can_shift && Stage::Type::SHIFT != (0, 0) {
+            return Err(Error::PipelineShiftAfterExpand(stage.name()));
+        }
+        if Stage::Type::TYPE == RenderPipelineStageType::Extend {
+            self.can_shift = false;
+        }
+        self.pipeline.stages.push(Box::new(stage));
+        Ok(self)
+    }
+
+    fn build(mut self) -> Result<Self::RenderPipeline> {
+        let channel_info = &mut self.pipeline.channel_info;
+        for stage in (0..self.pipeline.stages.len()).rev() {
+            for chan in 0..channel_info[0].len() {
+                if channel_info[stage][chan].ty.is_none() {
+                    channel_info[stage][chan].ty = channel_info[stage + 1][chan].ty;
+                }
+                // Arithmetic overflows here should be very uncommon, so custom error variants
+                // are probably unwarranted.
+                channel_info[stage][chan].downsample.0 = channel_info[stage][chan]
```

Comment:
Would a reference to this tuple make sense? I feel it would make this slightly easier to follow.

Reply:
Made the code somewhat nicer